### DISCUSSION TITLE
webpack cache does not update when node_modules change, even after server restart

### DISCUSSION THREAD

**Discussion by gaurav5430** [vercel/next.js#34380] [Ideas]
### Run `next info` (available from version 12.0.8 and up)

NA

### What version of Next.js are you using?

11.1.3

### What version of Node.js are you using?

v12.22.8

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

own setup with AWS

### Describe the Bug

During development, we symlink to some local dependencies and install them in our nextjs project. These show up in node modules. When any file in these dependencies change, we expect that next js would hot reload the changes, instead what we have observed is that the webpack build runs again, but the changes are not reflected on the UI. 

On debugging more, we could figure out that although the files in node_modules have changed, the files loaded by webpack on the browser are the same and not the new changed ones. We tried restarting the server to see if it fixes this, that does not seem to work.  Since node modules is a special folder, we figured that it is possible that there are some specific rules for node modules due to which the changes are not being watched or something, so we also tried making some other changes in the app code so that hot reload happens and it can maybe pickup the changes in node modules as well. This also does not work.

On further understanding, it seems like the filesystem cache created by nextjs webpack config, is not updating even though the symlinked dependency has updated in node modules. If we delete the cache and run the server again, it works alright.

This was working fine before nextjs 11, and webpack 5

### Expected Behavior

We would expect the changes in node modules to be hot reloaded and included in the build 

### To Reproduce

- add a custom webpack config
- add config.resolve.symlinks = false in your custom config
- in package json refer to a package from your local filesystem (npm will treat it as a symlinked package)
- npm i to install the local package
- consume the local package in your app
- start the server
- things work as expected
- make some changes in the local package
- these changes are not reflected / not hot reloaded 
- try restarting the server - still do not get reflected
- try making other code changes in the app, the other code changes get reflected, but not the ones in local package
- try deleting the node modules , package lock, and do an npm install again  - things work fine, changes are available
- OR try deleting the .next/cache , restart the server - things work fine, changes are available
- make some changes in the webpack config, restart the server -  this sometimes leads to cache invalidation, things work fine, changes are available

**Comment by balazsorban44** [comment]
Hi, Webpack 5 introduced caching for faster builds, which Webpack 4 lacked.

`node_modules` are heavily cached for the best performance, and are only invalidated if your package manager sees a change through one of these files:

- `package.json`
- `yarn.lock`
- `package-lock.json`

A possible solution could be that you use `externalDir` and let Next.js transpile your dependencies in development: https://github.com/vercel/next.js/discussions/26420

This will add HMR to your packages as well, improving the feedback loop during development.

**Reply by AlanSl** [reply]
> node_modules are heavily cached for the best performance, and are only invalidated if your package manager sees a change through one of these files:
> 
> package.json
> yarn.lock
> package-lock.json

Is there any way to configure this, to add to the list of files that are watched and which invalidate the cache if changed?

For example: we use `yalc` for pre-publish testing of bundled packages outside their repo: `yalc publish` in repo A and `yalc add` or `yalc update` in repo B (which includes a NextJS app). Then repo B can test a realistic published bundle in a NextJS app before publishing to live registries, and spot bundling-related issues such as large size, broken treeshaking, implicit dependency/versioning errors, etc. Each yalc update updates a `yalc.lock` file where the dependency bundle has changed. 

If we could locally add `'yalc.lock'` to the above list in our NextJS app's webpack config somewhere, it'd probably all just work fine, but as is we have to manually wipe the caches each update.

**Reply by morhekil** [reply]
@AlanSl we ran into a similar issue, and it looks like cache invalidation can be done by adding a lock file to webpack's [`cache.buildDependencies`](https://webpack.js.org/configuration/cache/#cachebuilddependencies) object.

``` js
// this goes into next.config.js

module.exports = {
  webpack: (config) => {
    config.cache.buildDependencies.mydeps = ['/path/to/lockfile'];
    return config;
  }
};
```

**Reply by brandoncollins7** [reply]
This has made things very difficult for users of bit.dev components. Since bitdev creates symlinks into node_modules you effectly lose any hotreload capabilities when working on your bit components. Only option to view a small change in your app is to delete the .next folder and wait a year for your application to fully recompile:(:(

**Reply by pringelmann** [reply]
This answer on stackoverflow solved it for me: [link](https://stackoverflow.com/questions/71345874/should-changes-in-node-modules-invalidate-the-webpack-build-cache/72576675#72576675)

```js
snapshot: {
    managedPaths: [/^(.+?[\\/]node_modules)[\\/]((?!your-modules)).*[\\/]*/],
},
```

Related Webpack docs: https://webpack.js.org/configuration/other-options/#managedpaths

**Reply by gaurav5430** [reply]
@Pringels 
the above regular expression (after changing `your-modules` to my scoped module path) leads to this error in my nextjs config

```
ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
 - configuration[0].snapshot.managedPaths[0] should be a non-empty string.
   -> A path to a managed directory (usually a node_modules directory).
 - configuration[1].snapshot.managedPaths[0] should be a non-empty string.
   -> A path to a managed directory (usually a node_modules directory).
    at validate (.../node_modules/next/dist/compiled/schema-utils3/index.js:1:153657)
```

**Reply by gaurav5430** [reply]
@morhekil 

I don't get what you mean by adding 'a' lockfile 

Has anyone got this to work?

**Reply by aaronnuu** [reply]
> @morhekil
> 
> I don't get what you mean by adding 'a' lockfile
> 
> Has anyone got this to work?

All I really wanted was for webpack to not cache my node modules so I could test some changes that I was making so all I did was add a reference to a file that doesn't exist, which worked for me.
i.e. 
```js
module.exports = {
  webpack: (config) => {
    config.cache.buildDependencies.mydeps = ['/nonexistent/file/location'];
    return config;
  }
};
```

**Reply by pringelmann** [reply]
> @Pringels the above regular expression (after changing `your-modules` to my scoped module path) leads to this error in my nextjs config
> 
> ```
> ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
>  - configuration[0].snapshot.managedPaths[0] should be a non-empty string.
>    -> A path to a managed directory (usually a node_modules directory).
>  - configuration[1].snapshot.managedPaths[0] should be a non-empty string.
>    -> A path to a managed directory (usually a node_modules directory).
>     at validate (.../node_modules/next/dist/compiled/schema-utils3/index.js:1:153657)
> ```

@gaurav5430 If you are getting that error then your regex is not matching up with your module. You can maybe try test your regex separately. If it finds the path to your module the error should go away.

**Reply by gaurav5430** [reply]
> > @Pringels the above regular expression (after changing `your-modules` to my scoped module path) leads to this error in my nextjs config
> > 
> > ```
> > ValidationError: Invalid configuration object. Webpack has been initialized using a configuration object that does not match the API schema.
> >  - configuration[0].snapshot.managedPaths[0] should be a non-empty string.
> >    -> A path to a managed directory (usually a node_modules directory).
> >  - configuration[1].snapshot.managedPaths[0] should be a non-empty string.
> >    -> A path to a managed directory (usually a node_modules directory).
> >     at validate (.../node_modules/next/dist/compiled/schema-utils3/index.js:1:153657)
> > ```
> 
> @gaurav5430 If you are getting that error then your regex is not matching up with your module. You can maybe try test your regex separately. If it finds the path to your module the error should go away.

yeah, that makes sense, but what is the regex trying to do?

is it trying to mark that everything in node modules apart from my package is a managedPath
OR 
that my package in node_modules is a managedPath

**Reply by pringelmann** [reply]
Ah I see, sorry I misread the error. It looks like regex support was added in webpack `5.59.0`: 
https://github.com/webpack/webpack/releases/tag/v5.59.0

Commit:
https://github.com/webpack/webpack/commit/5e16d642ad721e785c8c2a7fde1f51efccc78613

What version of Next.js are you using?

**Reply by praveenpuglia** [reply]
Hey @Pringels 
I am from the same team as @gaurav5430 . We use webpack `^5.62.1`. The funny thing is with the `snapshot.managedPaths` settings, I do see that webpack compiles things again and logs in console. The web console also shows `[Fast Refresh] rebuilding` but no real change is reflected in the actual app. 

Is there something we are missing in the next configuration?

**Reply by praveenpuglia** [reply]
BTW, realized that next is using 5.51.X instead of the specified webpack version in our package.json. Got to know this by going to **next.config.js** and logging webpack version from here.
```js
webpack(config, { dev, isServer, webpack }) {
  console.log(webpack.version);
  ...
}
```

**Reply by gaurav5430** [reply]
This is how we were able to fix it in a non next project
https://github.com/webpack/webpack/issues/11612#issuecomment-1448208868

**Reply by shirakaba** [reply]
Here's how I've set up my `next.config.js`:

```diff
+ /**
+  * Flip this to `true` whenever you want to watch changes in
+  * node_modules (e.g. to debug a dependency).
+  * 
+  * It's expensive to watch node_modules all the time, so make
+  * sure to put it back to `false` once you're done debugging.
+  * 
+  * @see https://github.com/vercel/next.js/discussions/34380#discussioncomment-8936245
+  */
+ const watchNodeModules = false;
  
  /**
    * @type {import('next').NextConfig}
    * @see https://nextjs.org/docs/pages/api-reference/next-config-js
    */
  const nextConfig = {
    // ... unrelated stuff...

+   webpack: (config, { isServer, nextRuntime }) => {
+     if (watchNodeModules) {
+       // It's recommended to opt-out of managedPaths if editing files in
+       // node_modules directly.
+       // @see https://github.com/webpack/webpack/issues/11612#issuecomment-705790881
+       // @see https://webpack.js.org/configuration/other-options/#managedpaths
+       config.snapshot.managedPaths = [];
+
+       // config.watchOptions isn't directly writable, so we use defineProperty.
+       Object.defineProperty(config, "watchOptions", {
+         ...Object.getOwnPropertyDescriptor(config, "watchOptions"),
+         value: {
+           ...config.watchOptions,
+           ignored:
+             /^((?:[^/]*(?:\/|$))*)(\.(git|next))(\/((?:[^/]*(?:\/|$))*)(?:$|\/))?/,
+         },
+       });
+
+       // This seems to work without any changes to watchOptions.followSymlinks
+       // or resolve.symlinks, but in case it ever becomes relevant:
+       // @see https://github.com/webpack/webpack/issues/11612#issuecomment-1448208868
+     }
+ 
+     return config;
+   },
  };
  
  module.exports = nextConfig;
```

It's based on changing the original `config.watchOptions.ignored` to cut out the `node_modules` exclusion:

```diff
- /^((?:[^/]*(?:\/|$))*)(\.(git|next)|node_modules)(\/((?:[^/]*(?:\/|$))*)(?:$|\/))?/
+ /^((?:[^/]*(?:\/|$))*)(\.(git|next))(\/((?:[^/]*(?:\/|$))*)(?:$|\/))?/
```

Caution 1: If that Regex ever updates in a future version of Next.js, make sure to log `config.watchOptions.ignored` to identify its new value, then cut out the part of the rule that excludes `node_modules` again.

Caution 2: If you find that you are making changes and they're still not getting picked up, check that you're editing the right file. In my project, Webpack traverses only the MJS sources and disregards all CJS ones.

@balazsorban44 Quite honestly though, people have been asking for this feature [since 2017](https://github.com/vercel/next.js/issues/1721) so by now it should really be exposed somewhere as an easy-to-flip switch. There's no point saving developer time with lots of caching if they still lose the whole day restarting their dev server each time they want to add a single console.log somewhere.

**Reply by michaelhogg** [reply]
@shirakaba: Thank you so much, works perfectly! :tada:

**Reply by kuket15** [reply]
Thanks for sharing, it works perfectly also for me!
